eZ/Publish/Core/Repository/UserService.php

```diff
@@ -770,6 +748,108 @@ public function updateUser(APIUser $user, UserUpdateStruct $userUpdateStruct): A
         return $this->loadUser($loadedUser->id);
     }
 
+    /**
+     * Validates and updates just the user's password.
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\User $user
+     * @param string $newPassword
+     *
+     * @throws ContentFieldValidationException
+     * @throws MissingUserFieldTypeException
+     * @throws UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\ContentValidationException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    public function updateUserPassword(APIUser $user, string $newPassword): APIUser
```

Comment:
Where's the check that the old password is valid? Expected to see `string $oldPassword` here.

Reply:
IMO it's outside the scope of the user repository class if the developer checked the old password beforehand. In case where this method is called from - for example - password reset functionality, old password would not even be available.

From my point of view, it's the responsibility of the developer to check that kind of security. Even more so since it could not only be related to old password itself.

Reply:
> IMO it's outside the scope of the user repository class if the developer checked the old password beforehand. In case where this method is called from - for example - password reset functionality, old password would not even be available.
> 
> From my point of view, it's the responsibility of the developer to check that kind of security. Even more so since it could not only be related to old password itself.

But is it really used by reset password or is it just an idea? This is Public API. Public API provides security layer.

Reply:
Original `updateUser` method, while allowing the password to be changed, didn't require previous password to be passed, nor had any checks related to it.

At this point in time I password reset still uses `updateUser` method. I think it should also be adjusted once this is integrated into kernel, since both `PasswordChangeController` and `PasswordResetController` relied on the same set of roles and policies, and will be subject the same bug as reported (i.e. user unable to reset password despite having `user/password` policy).

The security check we're discussing here is not really related to user's password - it's about confirming that the logged in user is still the same person making the request. This can be done in more ways than just the old password check - for example by sending a text message with code to one's phone or sending an email with confirmation code. It's only that our eZPlatform meta package uses old password as a security mechanism here.
Limiting `updateUserPassword` to old password check only would make it a liability in the future IMO.

Reply:
> Original `updateUser` method, while allowing the password to be changed, didn't require previous password to be passed, nor had any checks related to it.

Hmm, this is a design inconsistent with what other API methods offer. However...

> The security check we're discussing here is not really related to user's password - it's about confirming that the logged in user is still the same person making the request (authentication), in contrast to what service does (authorization). This can be done in more ways than just the old password check - for example by sending a text message with code to one's phone or sending an email with confirmation code. It's only that our eZPlatform meta package uses old password as a security mechanism here.

This is a good point. It still should be supported by extension point so API is consistent, but this is more complicated topic, out of scope here.

+1